File: FMShield.cpp
#include "pxt.h"
#include "MicroBitPin.h"
#include "mbed.h"

#include "FMShield.h"
using namespace pxt;
r revInf;
s sendCmd;
SPI fmShield_spi(MOSI, MISO, SCK);
#define fmShield_SPI_Write_Byte(value) fmShield_spi.write(value)
DigitalOut NSS(MICROBIT_PIN_P16);
//% weight=2 color=#0050BB icon="\uf1b9"

namespace FMShield
{
//%
void sendCmdRevInf(uint8_t size)
{
    uint8_t i;
    //uint8_t size=128;
    fmShield_spi.format(8, 0);
    fmShield_spi.frequency(1000000);
    NSS = 0;

    for (i = 0; i < size; i++)
    {
        revInf.cInf[i] = fmShield_SPI_Write_Byte(sendCmd.cSend[i]);
    }
    NSS = 1;
}

/**
    * TODO: 在此处描述您的函数
    * @param m 在此处描述参数,
    * @param t 在此处描述参数, 
    * @param p 在此处描述参数
    */
    //% weight=13
    //% blockGap=8
    //% blockId=sensor_ev3GyroSensor block="EV3 陀螺仪|%m|模式|%mode"
    //% shim=FMShield::ev3UartSensor
int ev3UartSensor(uint8_t s, uint8_t mode)
{
    // Add code here
    //uint8_t startN = 4; //8byte
    float mm;
    //int intMm;
    sendCmd.cSend[0 + s * 4] = 1;
    sendCmd.cSend[2 + s * 4] = mode;
    sendCmdRevInf(128);
    switch (s)
    {
    case 0:
        mm = revInf.inf.s1.Actual;
        break;
    case 1:
        mm = revInf.inf.s2.Actual;
        break;
    case 2:
        mm = revInf.inf.s3.Actual;
        break;
    case 3:
        mm = revInf.inf.s4.Actual;
        break;
    case 4:
        mm = revInf.inf.s5.Actual;
        break;
    case 5:
        mm = revInf.inf.s6.Actual;
        break;
    }

    return int(mm);
}

/**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=16
    //% blockGap=8
    //% blockId=sensor_nxtLightSensor block="NXT 光电|%s"
    
int nxtLightSensor(uint8_t s)
{
    // Add code here
    //uint8_t startN = 4; //8byte

    int intMm;
    //sendCmd.cSend[0 + s * 4] = 1;
    //sendCmd.cSend[2 + s * 4] = mode;
    sendCmdRevInf(128);
    switch (s)
    {
    case 0:
        intMm = revInf.inf.adc.s1_ANA;
        break;
    case 1:
        intMm = revInf.inf.adc.s2_ANA;
        break;
    case 2:
        intMm = revInf.inf.adc.s3_ANA;
        break;
    case 3:
        intMm = revInf.inf.adc.s4_ANA;
        break;
    case 4:
        intMm = revInf.inf.adc.s5_ANA;
        break;
    case 5:
        intMm = revInf.inf.adc.s6_ANA;
        break;
    }

    return intMm;
}
//%
void setMotor(uint8_t m, uint8_t t, uint8_t p)
{
    // Add code here
    uint8_t startN = 24; // let startN : number = 24; //8byte
    uint8_t mm;          //let mm: number;

    uint8_t _mode = sendCmd.cSend[startN + m * 16]; //let _mode: number = cmd[sendCmd];
    mm = _mode & 0x3f;

    sendCmd.cSend[startN + m * 16] = mm + t * 128 + p * 64;
    //sendCmd.cSend[startN + m * 16] = 2;
}
//%
void setMotorMode(uint8_t m, uint8_t mode)
{
    // Add code here
    uint8_t startN = 24; //let startN: number = 24; //8byte
    uint8_t mm;          //    let mm : number;
    mm = sendCmd.cSend[startN + m * 16] & 0xc0;
    sendCmd.cSend[startN + m * 16] = mode | mm;
}
//%
void setMotorInversion(uint8_t m)
{
    // Add code here
    uint8_t startN = 25; //let startN : number = 25; //8byte
    sendCmd.cSend[startN + m * 16] = 1;
}
//%
void motorRun(uint8_t m, int speed)
{
    // Add code here
    uint8_t startN = 28; //let startN: number = 28; //8byte

    setMotorMode(m, MOTOR_FIXED_DRIVE);
    speed = speed * 20;
    if (speed > 2000)
    {
        speed = 2000;
    }
    if (speed < -2000)
    {
        speed = -2000;
    }
    sendCmd.cSend[startN + m * 16] = speed % 255;
    sendCmd.cSend[startN + 1 + m * 16] = speed >> 8;
    sendCmdRevInf(128);
}
//%
void motorFineRun(uint8_t m, int speed)
{
    // Add code here
    uint8_t startN = 28; //let startN: number = 28; //8byte

    setMotorMode(m, MOTOR_FIXED_DRIVE);
    if (speed > 2000)
    {
        speed = 2000;
    }
    if (speed < -2000)
    {
        speed = -2000;
    }
    sendCmd.cSend[startN + m * 16] = speed % 255;
    sendCmd.cSend[startN + 1 + m * 16] = speed >> 8;
    sendCmdRevInf(128);
}
//%
void motorConstantRun(uint8_t m, int speed)
{
    // Add code here
    uint8_t startN = 28; //let startN: number = 28; //8byte

    setMotorMode(m, MOTOR_PID_SPEED);
    if (speed > 100)
    {
        speed = 100;
    }
    if (speed < -100)
    {
        speed = -100;
    }
    sendCmd.cSend[startN + m * 16] = speed % 255;
    sendCmd.cSend[startN + 1 + m * 16] = speed >> 8;
    sendCmdRevInf(128);
}
/**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=25
    //% blockGap=8
    //% blockId=motor_motorReadPosition block="读取电机码盘数|%m"
    
int32_t motorReadPosition(uint8_t m)
{
    // Add code here
    uint8_t startN = 76; //let startN: number = 76; //8byte
    long mm;             //let mm: number;
    sendCmdRevInf(128);
    mm = revInf.cInf[startN + m * 12] + revInf.cInf[startN + 1 + m * 12] * 256 + revInf.cInf[startN + 2 + m * 12] * 256 * 256 + revInf.cInf[startN + 3 + m * 12] * 256 * 256 * 256;
    return mm;
}
//%
void motorRunPosition(uint8_t m, int speed, long position)
{
    // Add code here
    uint8_t startN = 28; //let startN: number = 28; //8byte

    long p = motorReadPosition(m); //let p: number = motorReadPosition(m);
    setMotorMode(m, MOTOR_PID_POSITION);
    if (speed > 100)
    {
        speed = 100;
    }
    if (speed < -100)
    {
        speed = -100;
    }
    sendCmd.cSend[startN + m * 16] = speed % 255;
    sendCmd.cSend[startN + 1 + m * 16] = speed >> 8;
    position = position + p;
    sendCmd.cSend[startN + 4 + m * 16] = position % 255;
    sendCmd.cSend[startN + 4 + 1 + m * 16] = (position >> 8) % 255;
    sendCmd.cSend[startN + 4 + 2 + m * 16] = (position >> 16) % 255;
    sendCmd.cSend[startN + 4 + 3 + m * 16] = (position >> 24) % 255;

    sendCmdRevInf(128);
}
//%
void motorGoToPosition(uint8_t m, int speed, int32_t position)
{
    uint8_t startN = 28; //let startN: number = 28; //8byte
    setMotorMode(m, MOTOR_PID_POSITION);
    if (speed > 100)
        speed = 100;
    if (speed < -100)
        speed = -100;

    sendCmd.cSend[startN + m * 16] = speed % 255;
    sendCmd.cSend[startN + 1 + m * 16] = speed >> 8;
    sendCmd.cSend[startN + 4 + m * 16] = position % 255;
    sendCmd.cSend[startN + 4 + 1 + m * 16] = (position >> 8) % 255;
    sendCmd.cSend[startN + 4 + 2 + m * 16] = (position >> 16) % 255;
    sendCmd.cSend[startN + 4 + 3 + m * 16] = (position >> 24) % 255;
    sendCmdRevInf(128);
}

} // namespace FMShield

File: FMShield.ts

/**
 * 使用此文件来定义自定义函数和图形块。
 * 想了解更详细的信息，请前往 https://makecode.microbit.org/blocks/custom
 */
enum Ev3Sensor {
    //% block=S2
    S2 = 1,
    //% block=S3
    S3 = 2,
    //% block=S4
    S4 = 3,
    //% block=S5
    S5 = 4,
    //% block=S6
    S6 = 5
}

enum NxtSensor {
    //% block=S1
    S1 = 0,
    //% block=S2
    S2 = 1,
    //% block=S3
    S3 = 2,
    //% block=S4
    S4 = 3,
    //% block=S5
    S5 = 4,
    //% block=S6
    S6 = 5
}

enum Ev3UltrasonicSensorMode {
    //% block=厘米
    Centimeters = 0,
    //% block=英寸
    Inches = 1,
    //% block=监听
    Listen = 2
}

enum NxtLightSensorMode {
    //% block=反射光
    Reflected = 0,
    //% block=环境光
    Ambient = 1
    
}

enum Ev3ColorSensorMode {
    //% block=反射光
    COL_REFLECT = 0,
    //% block=环境光
    COL_AMBIENT = 1,
    //% block=颜色
    COL_COLOR = 2,
    //% block=反射光原始值
    COL_REF_RAW = 3,
    //% block=RGB原始值
    COL_RGB_RAW = 4,
    //% block=校准
    COL_CAL = 5
}

enum Ev3GyroSensorMode {
    //% block=角度
    Angle = 0,
    //% block=角速度
    Rate = 1,
    //% block=快速
    Fast = 2,
    //% block=校准
    Calibration = 4
}

enum FMSensor {
    //% block=S1
    S1 = 0,
    //% block=S2
    S2 = 1,
    //% block=S3
    S3 = 2,
    //% block=S4
    S4 = 3,
    //% block=S5
    S5 = 4,
    //% block=S6
    S6 = 5
}



enum FMMotorType {
    //% block=FM25
    FM25 = 0,
    //% block=LEGO
    LEGO = 1
}
enum FMPowerType {

    //% block=9V
    POWER_9V = 0,
    //% block=12V
    POWER_12V = 1
}
enum FMMotorSN {

    //% block=M1
    M1 = 0,
    //% block=M2
    M2 = 1,
    //% block=M3
    M3 = 2,
    //% block=M4
    M4 = 3
}
enum FMMotorMode {

    //% block=COAST 
    COAST = 0,
    //% block=BRAKE
    BRAKE = 1,
    //% block=FIXED_DRIVE
    FIXED_DRIVE = 2,
    //% block=PID_POSITION
    PID_POSITION = 3,
    //% block=PID_SPEED
    PID_SPEED = 4
}
/**
 * 自定义图形块
 */
//% weight=2 color=#0050BB icon="\uf1b9"
namespace FMShield {
    
    
    /**
        * TODO: 在此处描述您的函数
        * @param m 在此处描述参数,
        * @param t 在此处描述参数, 
        * @param p 在此处描述参数
        */
    //% weight=16
    //% blockGap=8
    //% blockId=sensor_nxtLightSensor block="NXT 光电|%s"
    //% shim=FMShield::nxtLightSensor


    export function sensor_nxtLightSensor(s: NxtSensor): number {
        return;
    }


    /**
        * TODO: 在此处描述您的函数
        * @param m 在此处描述参数,
        * @param t 在此处描述参数, 
        * @param p 在此处描述参数
        */ 
    //% weight=15
    //% blockGap=8
    //% blockId=sensor_ev3UltrasonicSensor block="EV3 超声波|%s|模式|%mode"
    //% shim=FMShield::ev3UartSensor


    export function sensor_ev3UltrasonicSensor(s: Ev3Sensor, mode: Ev3UltrasonicSensorMode): number {
        return;
    }
    /**
    * TODO: 在此处描述您的函数
    * @param m 在此处描述参数,
    * @param t 在此处描述参数, 
    * @param p 在此处描述参数
    */
    //% weight=14
    //% blockGap=8
    //% blockId=sensor_ev3ColorSensor block="EV3 颜色|%m|模式|%mode"
    //% shim=FMShield::ev3UartSensor

    export function sensor_ev3ColorSensor(m: Ev3Sensor, mode: Ev3ColorSensorMode): number {
        return;
    }

    /**
    * TODO: 在此处描述您的函数
    * @param m 在此处描述参数,
    * @param t 在此处描述参数, 
    * @param p 在此处描述参数
    */
    //% weight=13
    //% blockGap=8
    //% blockId=sensor_ev3GyroSensor block="EV3 陀螺仪|%m|模式|%mode"
    //% shim=FMShield::ev3UartSensor

    export function sensor_ev3GyroSensor(m: Ev3Sensor, mode: Ev3GyroSensorMode): number {
        return;
    }

    /**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t motor type; eg: FMMotorType.LEGO
     * @param p 在此处描述参数
     */
    //% weight=50
    //% blockGap=8
    //% blockId=motor_setMotor block="设置电机|%m|类型|%t|电源|%p"
    //% shim=FMShield::setMotor
    export function motor_setMotor(m: FMMotorSN, t: FMMotorType, p: FMPowerType): void {
        // Add code here
        return;
    }

    

    /**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=48
    //% blockGap=8
    //% blockId=motor_setMotorInversion block="设置电机反转|%m"
    //% shim=FMShield::setMotorInversion
    export function motor_setMotorInversion(m: FMMotorSN): void {
        return;
    }

    /**
         * TODO: 在此处描述您的函数
         * @param m 在此处描述参数,
         * @param t 在此处描述参数, 
         * @param p 在此处描述参数
         */
    //% weight=30
    //% blockGap=8
    //% blockId=motor_motorRun block="电机运行|%m|速度|%speed"
    //% speed.min=-100 speed.max=100
    //% shim=FMShield::motorRun
    export function motor_motorRun(m: FMMotorSN, speed: number): void {
        // Add code here
        return;
    }

    /**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=29
    //% blockGap=8
    //% blockId=motor_motorFineRun block="电机高精度运行|%m|速度|%speed"
    //% speed.min=-2000 speed.max=2000
    //% shim=FMShield::motorFineRun
    export function motor_motorFineRun(m: FMMotorSN, speed: number): void {
        // Add code here
        return;
    }

    /**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=28
    //% blockGap=8
    //% blockId=motor_motorConstantRun block="电机恒速运行|%m|速度|%speed"
    //% speed.min=-100 speed.max=100
    //% shim=FMShield::motorConstantRun
    export function motor_motorConstantRun(m: FMMotorSN, speed: number): void {
        return;
    }
    /**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=26
    //% blockGap=8
    //% blockId=motor_motorGoToPosition block="电机运行到|%m|速度|%speed|码盘数|%position"
    //% speed.min=-100 speed.max=100
    //% shim=FMShield::motorGoToPosition
    export function motor_motorGoToPosition(m: FMMotorSN, speed: number, position: number): void {
        return;
    }

/**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=27
    //% blockGap=8
    //% blockId=motor_motorRunPosition block="电机运行|%m|速度|%speed|码盘数|%position"
    //% speed.min=-100 speed.max=100
    //% shim=FMShield::motorRunPosition
    export function motor_motorRunPosition(m: FMMotorSN, speed: number, position: number): void {
        return;
    }

    /**
     * TODO: 在此处描述您的函数
     * @param m 在此处描述参数,
     * @param t 在此处描述参数, 
     * @param p 在此处描述参数
     */
    //% weight=25
    //% blockGap=8
    //% blockId=motor_motorReadPosition block="读取电机码盘数|%m"
    //% shim=FMShield::motorReadPosition
    export function motor_motorReadPosition(m: FMMotorSN): number {
       
        return ;
    }
}

